tls/s2n_connection_serialize.c

```diff
@@ -211,6 +216,62 @@ static S2N_RESULT s2n_connection_deserialize_parse(uint8_t *buffer, uint32_t buf
     return S2N_RESULT_OK;
 }
 
+/* Boringssl and AWS-LC do a special check in tls13 during the first call to encrypt after
+ * initialization. In the first call they assume that the sequence number will be 0, and therefore
+ * the provided nonce is equivalent to the implicit IV because 0 ^ iv = iv. The recovered implicit IV
+ * is stored and used later on to ensure the monotonicity of sequence numbers.
+ *
+ * In the case of deserialization, in the first call the sequence number may not be 0.
+ * Therefore the provided nonce cannot be considered to be the implicit IV because n ^ iv != iv.
+ * This inability to get the correct implicit IV causes issues with encryption later on.
+ *
+ * To resolve this we preform one throwaway encryption call with a zero sequence number after
+ * deserialization. This allows the libcrypto to recover the implicit IV correctly.
+ */
+static S2N_RESULT s2n_initialize_implicit_iv(struct s2n_connection *conn, struct s2n_connection_deserialize *parsed_values)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(parsed_values);
+
+#ifndef S2N_LIBCRYPTO_SUPPORTS_EVP_AEAD_TLS
+    return S2N_RESULT_OK;
+#endif
```

Comment:
Will this cause the compiler to error if S2N_LIBCRYPTO_SUPPORTS_EVP_AEAD_TLS is set, because of unreachable code? I'm not sure on the ordering of those operations.

Also, why do we need conditional compilation here if this compiles just fine regardless of support?

Reply:
We need a way to gate this code if we're not AWS-LC or Boringssl. @camshaft suggested a try-compile on the existence of the EVP_AEAD_TLS APIs, what would you prefer?

Reply:
For implementations that don't track nonces we don't need to perform this work. I guess if we want that to be more clear we can put the check in a separate function

```c
static bool s2n_crypto_evp_requires_iv_init(void)
{
#ifdef S2N_LIBCRYPTO_SUPPORTS_EVP_AEAD_TLS
    return true;
#else
    return false;
#endif
}
```